### DISCUSSION TITLE
Generic constructors for regular classes and structures

### DISCUSSION THREAD

**Discussion by EldHasp** [dotnet/csharplang#5810] [Language Ideas]
On the one hand, a constructor is a method.
But on the other hand, you can set ordinary methods and generic methods at the same time in a class, but constructors cannot.
You have to make either a derived wrapper class or static factory methods.

```
public class Command : ICommand
{
    private readonly Action<object> execute;
    private readonly Func<object, bool> canExecute;

    public Command(Action<object> execute, Func<object, bool> canExecute)
    {
        this.execute = execute;
        this.canExecute = canExecute;
    }

    // Instead of this method
    public static Command Create<T>(Action<T> execute, Func<T, bool> canExecute)
    {
        // Some code

        return command;
    }

    //  Would can be used such constructor
    public Command<T>(Action<T> execute, Func<T, bool> canExecute)
    {
        // Some code
    }
}
```

**Comment by jnm2** [comment]
It would be extremely confusing to see the syntax `new Foo<T>(...)` unless the type being constructed is actually `Foo<T>`.

What if the type is already generic *and* the constructor has generic type parameters like a regular method? Does that look like:
`new Foo<X><Y>(...)`

What is gained by allowing this? The example doesn't help me understand a use case.

**Reply by EldHasp** [reply]
An example that very often I have to implement through a factory method or a child class:

```

    var commObj = new Command(ExecuteObj, CanExecuteObj);
    var commInt = new Command(ExecuteInt, CanExecuteInt);

}

private void ExecuteObj(object obj) {/* Some Code*/ }
private bool CanExecuteObj(object obj) {/* Some Code*/ }
private void ExecuteInt(int num) {/* Some Code*/ }
private bool CanExecuteInt(int num) {/* Some Code*/ }


```
For a generic type (from your example), it is possible to specify two types: for a class and for a method: `new Foo<X,Y>(...)`.
A generic type for a method can be omitted if it can be determined from the method's parameters, as it is currently implemented for methods.

**Reply by jnm2** [reply]
I'd be very opposed to `new Foo<X, Y>` not giving me back an instance of a type that is `Foo<X, Y>`.

I'm not quite understanding the problem with static methods.

**Comment by Phantonia** [comment]
I think it would be better to add language support for factory methods, so that features like object initializers can be used on this generic `Create<T>` method rather than to extend the concept of constructors.

That would allow this:

```cs
public class Command : ICommand
{
    public string Name { get; init; }

    public static factory Command Create<T>(Action<T> execute, Func<T, bool> canExecute)
    {
        // Some code

        return new Command(...);
    }
}

Command command = Command.Create<int>(...) { Name = "My command" };
```

**Reply by EldHasp** [reply]
I agree with you, but this only partially solves the problem.
I would like a more complete solution.
Basically, a constructor is a method.
In a generic class, a method can have its own generic parameters, so let the constructor have them as well.

**Comment by HaloFour** [comment]
IMO it doesn't make a lot of sense for the constructor to be generic but for the type to be non-generic, that the use cases would be niche and better solved through existing means.
